src/ctelnet.cpp

```diff
@@ -928,6 +940,601 @@ std::tuple<QString, int, bool> cTelnet::getConnectionInfo() const
+        } else {
+            qDebug() << "WE send NEW_ENVIRON USERVAR" << it.key() << "as an empty VAL";
+        }
+    }
+}
+
+void cTelnet::appendNewEnvironValue(std::string &output, const QString &var, const bool isUserVar, const QMap<QString, QPair<bool, QString>> &newEnvironDataMap)
+{
+    if (newEnvironDataMap.contains(var)) {
+        // QPair first: NEW_ENVIRON_USERVAR indicator, second: data
+        const QPair<bool, QString> newEnvironData = newEnvironDataMap.value(var);
+        const QString val = newEnvironData.second;
+
+        if (newEnvironData.first != isUserVar) {
+            // RFC 1572: If a "type" is not followed by a VALUE (e.g., by another VAR,
+            // USERVAR, or IAC SE) then that variable is undefined.
+            output += isUserVar ? NEW_ENVIRON_USERVAR : NEW_ENVIRON_VAR;
+            output += escapeAndEncodeNewEnvironData(var).toStdString();
+            newEnvironVariablesSent.insert(var);
+
+            if (!isUserVar) {
+                qDebug() << "WE send NEW_ENVIRON VAR" << var << "with no VAL because we don't maintain it as VAR (use USERVAR!)";
+            } else {
+                qDebug() << "WE send NEW_ENVIRON USERVAR" << var << "with no VAL because we don't maintain it as USERVAR (use VAR!)";
+            }
+        } else {
+            output += isUserVar ? NEW_ENVIRON_USERVAR : NEW_ENVIRON_VAR;
+            output += escapeAndEncodeNewEnvironData(var).toStdString();
+            newEnvironVariablesSent.insert(var);
+            output += NEW_ENVIRON_VAL;
+
+            // RFC 1572: If a VALUE is immediately followed by a "type" or IAC, then the
+            // variable is defined, but has no value.
+            if (!val.isEmpty()) {
+                output += escapeAndEncodeNewEnvironData(val).toStdString();
+
+                if (!isUserVar) {
+                    qDebug() << "WE send NEW_ENVIRON VAR" << var << "VAL" << val;
+                } else {
+                    qDebug() << "WE send NEW_ENVIRON USERVAR" << var << "VAL" << val;
+                }
+            } else if (!isUserVar) {
+                qDebug() << "WE send NEW_ENVIRON VAR" << var << "as an empty VAL";
+            } else {
+                qDebug() << "WE send NEW_ENVIRON USERVAR" << var << "as an empty VAL";
+            }
+        }
+    } else {
+        // RFC 1572: If a "type" is not followed by a VALUE (e.g., by another VAR,
+        // USERVAR, or IAC SE) then that variable is undefined.
+        output += isUserVar ? NEW_ENVIRON_USERVAR : NEW_ENVIRON_VAR;
+        output += escapeAndEncodeNewEnvironData(var).toStdString();
+
+        if (!isUserVar) {
+            qDebug() << "WE send NEW_ENVIRON VAR" << var << "with no VAL because we don't maintain it";
+        } else {
+            qDebug() << "WE send NEW_ENVIRON USERVAR" << var << "with no VAL because we don't maintain it";
+        }
+    }
+}
+
+void cTelnet::sendIsNewEnvironValues(const QByteArray& payload)
```

Comment:
You can change it to sendNewEnvironValues. Please also change sendMNESValues.

There are two types of SENDs, an initial SEND IS (one for MNES, one for NEW_ENVIRON) and a SEND INFO (able to have a single method). So now we'd have a send and a sendInfo, instead of a sendIs and and sendInfo, which is not a big deal.